File: backend/src/main/java/com/maelrltt/norna/service/ProjectService.java
package com.maelrltt.norna.service;

import com.maelrltt.norna.dto.board.BoardResponse;
import com.maelrltt.norna.dto.project.ProjectBoardsUpdateRequest;
import com.maelrltt.norna.dto.project.ProjectNameUpdateRequest;
import com.maelrltt.norna.dto.project.ProjectRequest;
import com.maelrltt.norna.dto.project.ProjectResponse;
import com.maelrltt.norna.entity.*;
import com.maelrltt.norna.exception.ResourceNotFoundException;
import com.maelrltt.norna.repository.BoardRepository;
import com.maelrltt.norna.repository.ProjectRepository;
import com.maelrltt.norna.repository.SpaceRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ProjectService {
    private final ProjectRepository projectRepository;
    private final AuthService authService;
    private final SpaceMemberService spaceMemberService;
    private final SpaceRepository spaceRepository;
    private final BoardRepository boardRepository;

    @Transactional
    public ProjectResponse createProject(
            Authentication authentication,
            UUID spaceId,
            ProjectRequest projectRequest
    ) {
        User currentUser = authService.getCurrentUser(authentication);

        spaceMemberService.checkRole(spaceId, currentUser.getId(), SpaceRole.OWNER);

        Space space = spaceRepository.getSpaceById(spaceId);

        Project project = Project.builder()
                .name(projectRequest.name())
                .space(space)
                .build();
        projectRepository.save(project);

        List<Board> boards = boardRepository.findAllById(projectRequest.boardsId());
        for (Board board : boards) {
            if (!board.getSpace().getId().equals(spaceId)) {
                throw new ResourceNotFoundException("Board not found in this space");
            }
            board.setProject(project);
        }
        boardRepository.saveAll(boards);

        return new ProjectResponse(project.getId(), project.getName(),
                boards.stream().map(b -> new BoardResponse(b.getId(), b.getName(), b.getDescription())).toList());
    }

    public List<ProjectResponse> getProjects(
            Authentication authentication,
            UUID spaceId
    ) {
        User currentUser = authService.getCurrentUser(authentication);

        spaceMemberService.checkRole(
                spaceId,
                currentUser.getId(),
                SpaceRole.MEMBER,
                SpaceRole.OWNER
        );

        List<Project>  projects = projectRepository.findAllBySpaceId(spaceId);

        return projects.stream()
                .map(project -> new ProjectResponse(
                        project.getId(),
                        project.getName(),
                        project.getBoards().stream()
                                .map(b -> new BoardResponse(b.getId(), b.getName(), b.getDescription()))
                                .toList()
                )).toList();
    }

    @Transactional
    public ProjectResponse updateProjectName(
            Authentication authentication,
            UUID spaceId,
            UUID projectId,
            ProjectNameUpdateRequest projectNameUpdateRequest
    ) {
        User currentUser = authService.getCurrentUser(authentication);

        spaceMemberService.checkRole(spaceId, currentUser.getId(), SpaceRole.OWNER);

        Project project = projectRepository
                .findByIdAndSpaceId(projectId, spaceId)
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        project.setName(projectNameUpdateRequest.name());

        return new ProjectResponse(
                        project.getId(),
                        project.getName(),
                        project.getBoards().stream()
                                .map(b -> new BoardResponse(b.getId(), b.getName(), b.getDescription()))
                                .toList()
                );
    }

    @Transactional
    public ProjectResponse updateProjectBoards(
            Authentication authentication,
            UUID spaceId,
            UUID projectId,
            ProjectBoardsUpdateRequest projectBoardsUpdateRequest
    ) {
        User currentUser = authService.getCurrentUser(authentication);

        spaceMemberService.checkRole(spaceId, currentUser.getId(), SpaceRole.OWNER);

        Project project = projectRepository
                .findByIdAndSpaceId(projectId, spaceId)
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        List<Board> newBoards = boardRepository.findAllById(projectBoardsUpdateRequest.boardIds());

        if (newBoards.size() != projectBoardsUpdateRequest.boardIds().size()) {
            throw new ResourceNotFoundException("One or more boards not found");
        }

        for (Board board : newBoards) {
            if (!board.getSpace().getId().equals(spaceId)) {
                throw new ResourceNotFoundException("Board not found in this space");
            }
        }

        // Remove boards
        List<Board> currentBoards = boardRepository.findAllByProjectId(projectId);
        for (Board board : currentBoards) {
            if (!newBoards.contains(board)) {
                board.setProject(null);
            }
        }

        // Add new boards to the project
        for (Board board : newBoards) {
            board.setProject(project);
        }

        boardRepository.saveAll(currentBoards);
        boardRepository.saveAll(newBoards);

        return new ProjectResponse(
                project.getId(),
                project.getName(),
                newBoards.stream()
                        .map(b -> new BoardResponse(b.getId(), b.getName(), b.getDescription()))
                        .toList()
        );
    }

    @Transactional
    public void deleteProject(
            Authentication authentication,
            UUID spaceId,
            UUID projectId
    ) {
        User currentUser = authService.getCurrentUser(authentication);

        spaceMemberService.checkRole(spaceId, currentUser.getId(), SpaceRole.OWNER);

        Project project = projectRepository
                .findByIdAndSpaceId(projectId, spaceId)
                .orElseThrow(() -> new ResourceNotFoundException("Project not found"));

        // Remove boards
        List<Board> boards = boardRepository.findAllByProjectId(projectId);
        for (Board board : boards) {
            board.setProject(null);
        }
        boardRepository.saveAll(boards);

        projectRepository.delete(project);
    }
}

File: frontend/src/app/service/project/project.service.ts
import { HttpClient, httpResource } from '@angular/common/http';
import { Service, computed, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ProjectBoardsUpdateRequest, ProjectNameUpdateRequest, ProjectRequest, ProjectResponse } from '../../models/project';
import { environment } from '../../../environments/environment';
import { BoardService } from '../board/board.service';
import { SpaceService } from '../space/space.service';

@Service()
export class ProjectService {
  private http = inject(HttpClient);
  private spaceService = inject(SpaceService);
  private boardService = inject(BoardService);

  private projectsResource = httpResource<ProjectResponse[]>(() => {
    const space = this.spaceService.actualSpace();

    if ((!this.spaceService.isReady() || !space) && (!this.boardService.isReady() || !this.boardService.boards ) ) {
      return undefined;
    }

    return `${environment.apiUrl}/space/${space.spaceId}/project`;
  });

  readonly projects = computed(() => this.projectsResource.value() ?? []);

  readonly isLoading = this.projectsResource.isLoading;

  readonly error = this.projectsResource.error;

  readonly isReady = computed(
    () => !this.isLoading() && !this.error() && this.projectsResource.value() !== undefined,
  );

  refresh() {
    this.projectsResource.reload();
  }

  getProjects(spaceId: string): Observable<ProjectResponse[]> {
    return this.http.get<ProjectResponse[]>(
      `${environment.apiUrl}/space/${spaceId}/project`,
      {
        withCredentials: true,
      }
    );
  }

  createProject(
    spaceId: string,
    projectRequest: ProjectRequest
  ): Observable<ProjectResponse> {
    return this.http.post<ProjectResponse> (
      `${environment.apiUrl}/space/${spaceId}/project`,
      projectRequest,
      {
        withCredentials: true,
      }
    );
  }

  updateProjectName(
    spaceId: string,
    projectId: string,
    projectNameUpdateRequest: ProjectNameUpdateRequest
  ): Observable<ProjectResponse> {
    return this.http.patch<ProjectResponse>(
      `${environment.apiUrl}/space/${spaceId}/project/${projectId}`,
      projectNameUpdateRequest,
      {
        withCredentials: true
      }
    )
  }

  updateProjectBoards(
    spaceId: string,
    projectId: string,
    projectBoardsUpdateRequest: ProjectBoardsUpdateRequest
  ): Observable<ProjectResponse> {
    return this.http.patch<ProjectResponse>(
      `${environment.apiUrl}/space/${spaceId}/project/${projectId}`,
      projectBoardsUpdateRequest,
      {
        withCredentials: true
      }
    )
  }

  deleteProject(
    spaceId: string,
    projectId: string
  ): Observable<void> {
    return this.http.delete<void>(
      `${environment.apiUrl}/space/${spaceId}/project/${projectId}`,
      {
        withCredentials: true
      }
    )
  }
}
